### PR TITLE
Add mStable save and swap

### PR DESCRIPTION
**Pull request by dimsome**
Protocol: https://www.mstable.org/
Docs: https://docs.mstable.org/ (user focused) and https://developers.mstable.org/ (developer focused)

This connector adds support for interacting with mStable. Polygon and Mainnet

Users can:
### Interact with mStable Save Vault:
- Deposit/Withdraw mUSD
- Deposit/Withdraw bAssets 
- Deposit/Withdraw fAsset, via Feeder Pool
- Claim Accrued Rewards in MTA

### Swap Stables
- bAsset to mUSD via mint 
- bAsset to bAsset via swap 
- bAsset to fAsset via Feeder Pool 
- mUSD to bAsset via redeem 
- mUSD to fAsset via Feeder Pool 
- fAsset to mUSD via Feeder Pool 
- fAsset to bAsset via Feeder Pool 

### Reference
- bAssets on mainnet: sUSD, DAI, USDT, USDC
- bAssets on Polygon: DAI, USDT, USDC
- fAsset = Feeder Pool assets e.g. alUSD, FEI on Mainnet, FRAX on Polygon

- [x] Review and approve connector @thrilok209 
- [x] Deploy and comment address
- [ ] Build and deploy resolver
- [ ] Add to DOCS

### CODE REVIEW

File: contracts/mainnet/connectors/mstable/main.sol [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+			mintedAmount = IMasset(mUsdToken).mint(
+				_token,
+				_amount,
+				_minOut,
+				address(this)
+			);
+		} else {
+			require(mintedAmount >= _minOut, "mintedAmount < _minOut");
+		}
+
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol:67]: Hey I think this should emit the event 
```solidity
event LogDeposit(address token, uint256 amount, address path);
```
like - 
```suggestion
_eventName = "logDeposit(address,uint256,address)";
_eventParam = abi.encode(_token,_amount, path);
```

Also where is this path address variable?

File: contracts/mainnet/connectors/mstable/main.sol [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+		);
+
+		approve(TokenInterface(_token), _path, _amount);
+		uint256 mintedAmount = IFeederPool(_path).swap(
+			_token,
+			mUsdToken,
+			_amount,
+			_minOut,
+			address(this)
+		);
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol:122]: Here, the event should be emitted!

File: contracts/mainnet/connectors/mstable/main.sol [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+				_token,
+				amountWithdrawn,
+				_minOut,
+				address(this)
+			);
+		} else {
+			require(amountWithdrawn >= _minOut, "amountWithdrawn < _minOut");
+		}
+
+		_eventName = "LogWithdraw()";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]:
```suggestion
		_eventName = "LogWithdraw(mUsdToken, amountWithdrawn, imUsdToken)";
```

File: contracts/mainnet/connectors/mstable/main.sol [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+		approve(TokenInterface(mUsdToken), _path, amountWithdrawn);
+		uint256 amountRedeemed = IFeederPool(_path).swap(
+			mUsdToken,
+			_token,
+			amountWithdrawn,
+			_minOut,
+			address(this)
+		);
+
+		_eventName = "LogRedeem()";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]: Can you fill this _event with correct params

File: contracts/mainnet/connectors/mstable/main.sol [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+		address rewardToken = _getRewardTokens();
+		uint256 rewardAmount = _getRewardInternalBal(rewardToken);
+
+		IBoostedSavingsVault(imUsdVault).claimReward();
+
+		uint256 rewardAmountUpdated = _getRewardInternalBal(rewardToken);
+
+		uint256 claimedRewardToken = sub(rewardAmountUpdated, rewardAmount);
+
+		_eventName = "LogClaimRewards()";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]: This too

File: contracts/mainnet/connectors/mstable/main.sol [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+			amountSwapped = IMasset(mUsdToken).swap(
+				_input,
+				_output,
+				_amount,
+				_minOut,
+				address(this)
+			);
+		}
+
+		_eventName = "LogSwap()";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]: ?

File: contracts/mainnet/connectors/mstable/main.sol [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+		// swaps mUSD to fAsset via Feeder Pool
+		amountSwapped = IFeederPool(_path).swap(
+			_input,
+			_output,
+			_amount,
+			_minOut,
+			address(this)
+		);
+
+		_eventName = "LogSwap()";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]: ?

File: contracts/mainnet/connectors/mstable/main.sol [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+		uint256 credits = ISavingsContractV2(imUsdToken).depositSavings(
+			_amount
+		);
+
+		// 2. Stake imUSD to Vault
+		approve(TokenInterface(imUsdToken), imUsdVault, credits);
+		IBoostedSavingsVault(imUsdVault).stake(credits);
+
+		// 3. Log Events
+		_eventName = "LogDeposit()";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]: ?

File: contracts/mainnet/connectors/mstable/main.sol [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+		internal
+		view
+		returns (uint256 a)
+	{
+		a = TokenInterface(_rewardToken).balanceOf(address(this));
+	}
+}
+
+contract ConnectV2mStable is mStableResolver {
+	string public constant name = "mStable-Mainnet-Connector-v1";
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]:
```suggestion
	string public constant name = "mStable-v1.0";
```

File: contracts/mainnet/connectors/mstable/main.sol (hunk 2 above)
Comment (dimsome) [reply contracts/mainnet/connectors/mstable/main.sol:122]: the _deposit function returns the event data. Does that work?

File: contracts/mainnet/connectors/mstable/main.sol (hunk 1 above)
Comment (dimsome) [reply contracts/mainnet/connectors/mstable/main.sol:67]: the `_deposit` function returns the event data, does it work? The path is to distinguish the event from where it was deposited:

- path feeder pool address => deposited from feeder pool
- path mUSD address => minted mUSD first
- path imUSD address => just deposited in imUSD

Does it make sense or it is unnecessary?

Comment (pradyuman-verma) [reply contracts/mainnet/connectors/mstable/main.sol:67]: Make sense, just make sure the event data is proper.

File: contracts/mainnet/connectors/mstable/main.sol (hunk 2 above)
Comment (pradyuman-verma) [reply contracts/mainnet/connectors/mstable/main.sol:122]: Totally

**Comment by pradyuman-verma** [comment]
@dimsome Can you attach a screenshot of test cases?
Just wanna check everything is working fine!

**Comment by dimsome** [comment]
## Mainnet:
Screen Shot 2022-01-20 at 13 57 02
Screen Shot 2022-01-20 at 13 57 24

## Polygon:
Screen Shot 2022-01-20 at 14 07 13
Screen Shot 2022-01-20 at 14 07 34

File: contracts/mainnet/connectors/mstable/main.sol [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+	 * @dev Deposit to Save from any asset
+	 * @notice Called internally from deposit functions
+	 * @param _token Address of token to deposit
+	 * @param _amount Amount of token to deposit
+	 * @param _path Path to mint mUSD (only needed for Feeder Pool)
+	 * @return _eventName Event name
+	 * @return _eventParam Event parameters
+	 */
+
+	function _deposit(
```
Comment (pradyuman-verma) [inline contracts/mainnet/connectors/mstable/main.sol]: Why don't we move `_deposit`, `_withdraw`, `_getRewardToken`, `_getRewardInternalBal`  this functions to `helpers.sol` ?
I think these are just helper functions!

File: contracts/polygon/connectors/mstable/main.sol [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
+	 * @dev Deposit to Save from any asset
+	 * @notice Called internally from deposit functions
+	 * @param _token Address of token to deposit
+	 * @param _amount Amount of token to deposit
+	 * @param _path Path to mint mUSD (only needed for Feeder Pool)
+	 * @return _eventName Event name
+	 * @return _eventParam Event parameters
+	 */
+
+	function _deposit(
```
Comment (pradyuman-verma) [inline contracts/polygon/connectors/mstable/main.sol]: Same here, Let's move these internal functions in `helpers.sol`

File: test/mainnet/mstable/mstable.helpers.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
+import hre, { ethers } from "hardhat";
+import { IERC20Minimal__factory } from "../../../typechain";
+import { BigNumber as BN } from "ethers";
+
+export const DEAD_ADDRESS = "0x0000000000000000000000000000000000000001";
+export const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
```
Comment (pradyuman-verma) [inline test/mainnet/mstable/mstable.helpers.ts]:
```suggestion
export const ZERO_ADDRESS = ethers.constants.AddressZero;
```

File: test/polygon/mstable/mstable.helpers.ts [hunk 13]
```diff
@@ -0,0 +1,159 @@
+import hre, { ethers } from "hardhat";
+import { IERC20Minimal__factory } from "../../../typechain";
+import { BigNumber as BN } from "ethers";
+
+export const DEAD_ADDRESS = "0x0000000000000000000000000000000000000001";
+export const ZERO_ADDRESS = "0x0000000000000000000000000000000000000000";
```
Comment (pradyuman-verma) [inline test/polygon/mstable/mstable.helpers.ts]:
```suggestion
export const ZERO_ADDRESS = ethers.constants.AddressZero;
```

**Comment by dimsome** [comment]
made the adjustments and ran the tests again to ensure everything still works fine

**Comment by pradyuman-verma** [comment]
@dimsome, I can't find the usage of `getUint` and `setUint` in the `main.sol`.

```
Use getUint() (declared in Stores) for getting the value saved from the previous spell.
Use setUint() (declared in Stores) for setting the value to save for the future spell.
```

**Comment by pradyuman-verma** [comment]
Also, can you use `uint256(-1)` for the maximum amount transfer?

```
Use uint(-1) of type(uint256).max for the maximum amount everywhere
```

**Comment by pradyuman-verma** [comment]
Here is more info on connectors - https://github.com/Instadapp/dsa-connectors#new-connector-should-follow-the-current-directory-structure

**Comment by dimsome** [comment]
> Also, can you use `uint256(-1)` for the maximum amount transfer?
> 
> ```
> Use uint(-1) of type(uint256).max for the maximum amount everywhere
> ```

I don't use the max amount anywhere as far as I know. The ERC20 approve only approves for the to be transferred amount

**Comment by pradyuman-verma** [comment]
> 

But what if the user doesn't know his balance and wants to transfer the whole amount?
Don't you think this has to be the case?

**Comment by pradyuman-verma** [comment]
@dimsome
- Each external function in the `main.sol` should be payable.
- In the `swapViaFeeder` function, is the `_path` address is changing? Or Can we keep it hardcoded?
-

**Comment by dimsome** [comment]
@pradyuman-verma 
added payable

The _path is changing depends on the asset. RAI goes via the mUSD/RAI feeder pool, GUSD via the mUSD/GUSD Feeder Pool. Could do mapping, but the question is whether we could then add more assets later. Maybe it's easier the frontend is aware of what assets goes which path?

**Comment by pradyuman-verma** [comment]
> @pradyuman-verma added payable
> 
> The _path is changing depends on the asset. RAI goes via the mUSD/RAI feeder pool, GUSD via the mUSD/GUSD Feeder Pool. Could do mapping, but the question is whether we could then add more assets later. Maybe it's easier the frontend is aware of what assets goes which path?

Ohh Okay, Got it.
Then creating a mapping doesn't make sense, @thrilok209 what do you think?

**Comment by pradyuman-verma** [comment]
> @pradyuman-verma added payable
> 
> The _path is changing depends on the asset. RAI goes via the mUSD/RAI feeder pool, GUSD via the mUSD/GUSD Feeder Pool. Could do mapping, but the question is whether we could then add more assets later. Maybe it's easier the frontend is aware of what assets goes which path?

Maybe, If you provide an endpoint that can give us the _path based on tokens.

**Comment by dimsome** [comment]
We don't have an API for resolving this. Would it be okay for the App to contain this info? I think the app would need to be aware of the possible swaps in any case.

## Mainnet
alUSD (`0xbc6da0fe9ad5f3b0d58160288917aa56653660e9`) -> `0x4eaa01974b6594c0ee62ffd7fee56cf11e6af936`
GUSD (`0x056fd409e1d7a124bd7017459dfea2f387b6d5cd`) -> `0x4fb30c5a3ac8e85bc32785518633303c4590752d`
BUSD -(`0x4fabb145d64652a948d72533023f6e7a623c7c53`) -> `0xfe842e95f8911dcc21c943a1daa4bd641a1381c6`
RAI -> (`0x03ab458634910aad20ef5f1c8ee96f1d6ac54919`) -> `0x36f944b7312eac89381bd78326df9c84691d8a5b`
FEI -> (`0x956F47F50A910163D8BF957Cf5846D573E7f87CA`) -> `0x2F1423D27f9B20058d9D1843E342726fDF985Eb4`

## Polygon
Frax -> (`0x104592a158490a9228070e0a8e5343b499e125d0`) -> `0xb30a907084ac8a0d25dddab4e364827406fd09f0`

**Comment by pradyuman-verma** [comment]
Hey @dimsome 
Can you deploy this connector and comment addresses?

**Review by pradyuman-verma — APPROVED** [review]

**Comment by dimsome** [comment]
Polygon: `0xcCa258c4Ba565A5fE4Fa2D8Af603f3211Cce79ee`
Mainnet: `0xd6FDF729EBFB44658d1E2f4Abe32D5818bc432f3`

Contract verification was a bit of a hassle, didn't work with the script. Not sure why, so did it manually.

**Comment by pradyuman-verma** [comment]
> Polygon: `0xcCa258c4Ba565A5fE4Fa2D8Af603f3211Cce79ee` Mainnet: `0xd6FDF729EBFB44658d1E2f4Abe32D5818bc432f3`
> 
> Contract verification was a bit of a hassle, didn't work with the script. Not sure why, so did it manually.

Awesome, We will add these to our SDK soon.
Also, is the script taking too much time, or have you got any errors?

**Comment by dimsome** [comment]
>  Also, is the script taking too much time, or have you got any errors?

No errors, the verify on Etherscan/Polygonscan was running endlessly.

**Comment by pradyuman-verma** [comment]
> > Also, is the script taking too much time, or have you got any errors?
> 
> No errors, the verify on Etherscan/Polygonscan was running endlessly.

Oh, got it will fix it soon.

**Comment by pradyuman-verma** [comment]
### Deployed on polygon:

**Addr:** [0xcCa258c4Ba565A5fE4Fa2D8Af603f3211Cce79ee](https://polygonscan.com/address/0xcCa258c4Ba565A5fE4Fa2D8Af603f3211Cce79ee#code)
**Name** `MSTABLE-A`
**Tx** [here](https://polygonscan.com/tx/0x9af081d7a48e01c69b3a983b6619a9945e9bf68840f1dc47a6ba2031e7780c32)


### Deployed on mainnet:

**Addr:** [0xd6FDF729EBFB44658d1E2f4Abe32D5818bc432f3](https://etherscan.io/address/0xd6FDF729EBFB44658d1E2f4Abe32D5818bc432f3)
**Name** `MSTABLE-A`
**Tx** [here](https://etherscan.io/tx/0x0cc4c49dfcb213831a84fbe69ff8f49cb76709d1fd785410b0009fd3fb097ac8#eventlog)

**Pull request merged**